Cryptography (#3199)
OWASP/owasp-mastg: weaknesses/MASVS-CRYPTO/MASWE-0024.md

---

> title: Improper use of Message Authentication Code (MAC)

Reviewer: ```suggestion
title: Improper Use of Message Authentication Code (MAC)
```

refs:
- https://developer.android.com/privacy-and-security/cryptography#deprecated-functionality
- https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-131Ar2.pdf
- https://csrc.nist.gov/pubs/sp/800/224/ipd
- https://datatracker.ietf.org/doc/html/rfc6151
- https://web.archive.org/web/20170810051504/http://www.tcs.hut.fi/old/papers/aura/aura-csfws97.pdf
- https://en.wikipedia.org/wiki/Replay_attack
draft:
  description: Improper use of MAC. E.g: allowing the user to control the input.
    may expose cryptographic vulnerabilities, affecting data integrity.
  topics: null
status: draft

---

> Improper use of MAC by e.g by generating a MAC over a message without the timestamp can make the application susceptible for replay attacks.

Reviewer 2: ```suggestion
Improper use of a MAC by e.g generating a MAC over a message without the timestamp can make the application susceptible for replay attacks.
```

> Another common issue is using HMAC with on low-entropy input with any type of general based hashing algorithm like MD5, SHA-1, SHA-2 or even SHA-3 on low-entropy input like user supplied passwords and pins. HMAC aren't designed for low-entropy inputs or low-entropy keys. Doing so will result in producing "weak" message digest that easily can be exploited.

Reviewer 2: ```suggestion
Another common issue is using an HMAC with any type of general based hashing algorithm like MD5, SHA-1, SHA-2 or even SHA-3 on low-entropy input like user supplied passwords and pins. HMAC aren't designed for low-entropy inputs or low-entropy keys. Doing so will result in producing "weak" message digests that easily can be exploited.
```
